### PR TITLE
WIP: Updating --help options to a sensible order

### PR DESCRIPTION
**Pull request by wsandin**
Pushed to get going. Will work on the remaining options tomorrow.

### CODE REVIEW

**Comment by wsandin** [comment]
WIP for #4459, awaiting feedback on the ordering before updating the OPTIONS doc.

File: src/options.c
Comment (solardiz) [inline src/options.c:271]: Why?! I think these two `--single-*` options were more appropriate in their old place. However, I'd move `--[no-]single-retest-guess` to after `--no-single-pair`.

File: src/options.c
Comment (solardiz) [inline src/options.c]: Inconsistent indentation.

**Review by solardiz — COMMENTED** [review]
Just some initial quick comments. This is hard to review trying to guess the rationale between the specific changes.

File: src/options.c
Comment (wsandin) [reply src/options.c]: Missed that one again, just started having a read in `doc/README.coding.style` and setup my project with indentation properly again. Thanks for pointing it out!

File: src/options.c
Comment (wsandin) [reply src/options.c:271]: No sensible rationale behind now that I look closer at it. Agreed it was more clear initially. I had the intention to come back and go through all the options tonight. 

Update in https://github.com/wsandin/john/commit/6235b2956bc740bacc267bbd3ed436ac61dd64b0

**Comment by wsandin** [comment]
> Just some initial quick comments. This is hard to review trying to guess the rationale between the specific changes.

@solardiz That's understandable! I'll groom through it a few times throughout the weekend. Also happy to hop on a call or similar if you want walk through it together. I guess there's no #openwall IRC channel or similar?

**Comment by solardiz** [comment]
@wsandin With the coordination this starts to require, it gets to a point where I feel we could prefer to switch to a different approach - e.g., maybe I and/or someone else would try and re-order lines in a text file containing the output of `--help` (and add comments e.g. beyond column 80 on why the specific re-orderings are suggested), and you'd merely need to put that into code.

> I guess there's no #openwall IRC channel or similar?

We used to have `#openwall` on Freenode IRC network, but personally I haven't been there in years and I doubt it'd be a productive way to work on this issue.

**Comment by wsandin** [comment]
> @wsandin With the coordination this starts to require, it gets to a point where I feel we could prefer to switch to a different approach - e.g., maybe I and/or someone else would try and re-order lines in a text file containing the output of `--help` (and add comments e.g. beyond column 80 on why the specific re-orderings are suggested), and you'd merely need to put that into code.
> 
> > I guess there's no #openwall IRC channel or similar?
> 
> We used to have `#openwall` on Freenode IRC network, but personally I haven't been there in years and I doubt it'd be a productive way to work on this issue.

@solardiz Sure - Sorry I couldn't be more of help on this issue. But I think you or someone else who worked on john for a longer period should have a much better idea of what makes sense in terms of standard and consistency.

**Review by magnumripper — APPROVED** [review]
For what its worth I don't see any obvious problem with this although I didn't check much (but a little) for things that was inadvertently dropped and things like that.

I like the style of "alphabetically - when not". That is, we first group the most important stuff and just sort the rest.

File: src/options.c
Comment (magnumripper) [inline src/options.c:283]: No wait, this line went astray! It's bound to be with the line before it.

Comment (wsandin) [reply src/options.c:283]: Just fixed in a new commit that I squashed.

**Comment by magnumripper** [comment]
#4618 will produce a conflict, at least unless we merge this one first.

**Review by ghost — CHANGES_REQUESTED** [review]
`--verbosity` appears twice.

**Comment by ghost** [comment]
Is it possible to sort --session, --status and --restore? BTW: --make-charset seems so sad and lonely where it is now.

```
--status[=NAME]            print status of a session [called NAME]
--restore[=NAME]           restore an interrupted session [called NAME]
--session=NAME             give a new session the NAME
--make-charset=FILE        make a charset file. It will be overwritten
```


* subformat appears twice. BTW: Me prefers:
```
--format=[NAME|CLASS][,..] force hash of type NAME. The supported formats can
                           be seen with --list=formats and --list=subformats.
                           Valid classes: dynamic, cpu, opencl, ztex, mask, omp,
                           all, enabled, disabled.
--subformat=FORMAT         pick a benchmark format for --format=crypt
claudio@home: $ _
```

**Comment by ghost** [comment]
All duplicates
```
      2 --config=FILE              use FILE instead of john.conf or john.ini
      2 --field-separator-char=C   use 'C' instead of the ':' in input and pot files
      2 --length=N                 shortcut for --min-len=N --max-len=N
      2 --log-stderr               log to screen instead of file
      2 --max-length=N             request a maximum candidate length in bytes
      2 --min-length=N             request a minimum candidate length in bytes
      2 --mkpc=N                   request a lower max. keys per crypt
      2 --no-loader-dupecheck      disable the dupe checking when loading hashes
      2 --no-log                   disables creation and writing to john.log file
      2 --subformat=FORMAT         pick a benchmark format for --format=crypt

```

**Comment by wsandin** [comment]
> All duplicates
> 
> ```
>       2 --config=FILE              use FILE instead of john.conf or john.ini
>       2 --field-separator-char=C   use 'C' instead of the ':' in input and pot files
>       2 --length=N                 shortcut for --min-len=N --max-len=N
>       2 --log-stderr               log to screen instead of file
>       2 --max-length=N             request a maximum candidate length in bytes
>       2 --min-length=N             request a minimum candidate length in bytes
>       2 --mkpc=N                   request a lower max. keys per crypt
>       2 --no-loader-dupecheck      disable the dupe checking when loading hashes
>       2 --no-log                   disables creation and writing to john.log file
>       2 --subformat=FORMAT         pick a benchmark format for --format=crypt
> ```


Err, this must have been something I didn't catch properly whilst rebasing. Just got home so will have a look now. I'll update with your other suggestions as well.

File: src/options.c
Comment (ghost) [inline src/options.c]: You added `fix-state-delay`. It is commented at line 163. 

Everything else is ok (I will add the `--help` output to the PR (so, if anyone wants to reorder things, he just need to use a text editor).

**Comment by ghost** [comment]
Final `--help` output (to hear from people). 

```
John the Ripper 1.9.0-jumbo-1+bleeding-d2701f3 2021-03-13 16:43:55 +0300 OMP [linux-gnu 64-bit x86_64 AVX2 AC]
Copyright (c) 1996-2021 by Solar Designer and others
Homepage: https://www.openwall.com/john/

Usage: john [OPTIONS] [PASSWORD-FILES]

--help                     print usage summary
--single[=SECTION[,..]]    "single crack" mode, using default or named rules
--single=:rule[,..]        same, using "immediate" rule(s)
--single-seed=WORD[,WORD]  add static seed word(s) for all salts in single mode
--single-wordlist=FILE     *short* wordlist with static seed words/morphemes
--single-user-seed=FILE    wordlist with seeds per username (user:password[s]
                           format)
--single-pair-max=N        override max. number of word pairs generated (6)
--no-single-pair           disable single word pair generation
--[no-]single-retest-guess override config for SingleRetestGuess
--wordlist[=FILE] --stdin  wordlist mode, read words from FILE or stdin
                  --pipe   like --stdin, but bulk reads, and allows rules
--rules[=SECTION[,..]]     enable word mangling rules (for wordlist or PRINCE
                           modes), using default or named rules
--rules=:rule[;..]]        same, using "immediate" rule(s)
--rules-stack=SECTION[,..] stacked rules, applied after regular rules or to
                           modes that otherwise don't support rules
--rules-stack=:rule[;..]   same, using "immediate" rule(s)
--rules-skip-nop           skip any NOP ":" rules (you already ran w/o rules)
--loopback[=FILE]          like --wordlist, but extract words from a .pot file
--mem-file-size=SIZE       size threshold for wordlist preload (default 2048 MB)
--dupe-suppression         suppress all dupes in wordlist (and force preload)
--incremental[=MODE]       "incremental" mode [using section MODE]
--incremental-charcount=N  override CharCount for incremental mode
--external=MODE            external mode or word filter
--mask[=MASK]              mask mode using MASK (or default from john.conf)
--markov[=OPTIONS]         "Markov" mode (see doc/MARKOV)
--mkv-stats=FILE           "markov" stats file (see doc/MARKOV)
--prince[=FILE]            PRINCE mode, read words from FILE
--prince-loopback[=FILE]   fetch words from a .pot file
--prince-elem-cnt-min=N    minimum number of elements per chain (1)
--prince-elem-cnt-max=[N]  maximum number of elements per chain (negative N is
                           relative to word length) (8)
--prince-skip=N            initial skip
--prince-limit=N           limit number of candidates generated
--prince-wl-dist-len       calculate length distribution from wordlist
--prince-wl-max=N          load only N words from input wordlist
--prince-case-permute      permute case of first letter
--prince-mmap              memory-map infile (not available with case permute)
--prince-keyspace          just show total keyspace that would be produced
                           (disregarding skip and limit)
--subsets[=CHARSET]        "subsets" mode (see doc/SUBSETS)
--subsets-required=N       The N first characters of "subsets" charset are
                           the "required set" (see doc/SUBSETS)
--subsets-min-diff=N       Minimum unique characters in subset
--subsets-max-diff=[-]N    Maximum unique characters in subset (negative N is
                           relative to word length)
--stdout[=LENGTH]          just output candidate passwords [cut at LENGTH]
--session=NAME             give a new session the NAME
--status[=NAME]            print status of a session [called NAME]
--[no-]crack-status        emit a status line whenever a password is cracked
--progress-every=N         emit a status line every N seconds
--restore[=NAME]           restore an interrupted session [called NAME]
--show[=left]              show cracked passwords [if =left, then uncracked]
--show=formats             show information about hashes in a file (JSON)
--show=invalid             show lines that are not valid for selected format(s)
--test[=TIME]              run tests and benchmarks for TIME seconds each
                           (if TIME is explicitly 0, test w/o benchmark)
--stress-test[=TIME]       loop self tests forever
--test-full=LEVEL          run more thorough self-tests
--no-mask                  used with --test for alternate benchmark w/o mask
--skip-self-tests          skip self tests
--users=[-]LOGIN|UID[,..]  [do not] load this (these) user(s) only
--groups=[-]GID[,..]       load users [not] of this (these) group(s) only
--shells=[-]SHELL[,..]     load users with[out] this (these) shell(s) only
--salts=[-]COUNT[:MAX]     load salts with[out] COUNT [to MAX] hashes, or
--salts=#M[-N]             load M [to N] most populated salts
--costs=[-]C[:M][,...]     load salts with[out] cost value Cn [to Mn]. For
                           tunable cost parameters, see doc/OPTIONS
--save-memory=LEVEL        enable memory saving, at LEVEL 1..3
--node=MIN[-MAX]/TOTAL     this node's number range out of TOTAL count
--log-stderr               log to screen instead of file
--verbosity=N              change verbosity (1-5 or 6 for debug, default 3)
--no-log                   disables creation and writing to john.log file
--bare-always-valid=Y      treat bare hashes as valid (Y/N)
--config=FILE              use FILE instead of john.conf or john.ini
--encoding=NAME            input encoding (eg. UTF-8, ISO-8859-1). See also
                           doc/ENCODINGS.
--input-encoding=NAME      input encoding (alias for --encoding)
--target-encoding=NAME     output encoding (used by format, see doc/ENCODINGS)
--force-tty                set up terminal for reading keystrokes even if we're
                           not the foreground process
--field-separator-char=C   use 'C' instead of the ':' in input and pot files
--fix-state-delay=N        performance tweak, see doc/OPTIONS
--internal-codepage=NAME   codepage used in rules/masks (see doc/ENCODINGS)
--[no-]keep-guessing       try finding plaintext collisions
--list=WHAT                list capabilities, see --list=help or doc/OPTIONS
--length=N                 shortcut for --min-len=N --max-len=N
--make-charset=FILE        make a charset file. It will be overwritten
--min-length=N             request a minimum candidate length in bytes
--max-length=N             request a maximum candidate length in bytes
--max-candidates=[-]N      gracefully exit after this many candidates tried.
                           (if negative, reset count on each crack)
--max-run-time=[-]N        gracefully exit after this many seconds (if negative,
                           reset timer on each crack)
--mkpc=N                   request a lower max. keys per crypt
--no-loader-dupecheck      disable the dupe checking when loading hashes
--pot=NAME                 pot file to use
--regen-lost-salts=N       brute force unknown salts (see doc/OPTIONS)
--reject-printable         reject printable binaries
--tune=HOW                 tuning options (auto/report/N)

OpenCL options:
--devices=N[,..]           set OpenCL device(s) (see --list=opencl-devices)
--force-scalar             force scalar mode
--force-vector-width=N     force vector width N
--lws=N                    force local worksize N
--gws=N                    force global worksize N

--fork=N                   fork N processes
--subformat=FORMAT         pick a benchmark format for --format=crypt
--format=[NAME|CLASS][,..] force hash of type NAME. The supported formats can
                           be seen with --list=formats and --list=subformats.
                           Valid classes: dynamic, cpu, opencl, ztex, mask, omp,
                           all, enabled, disabled.


```

**Comment by solardiz** [comment]
> Final `--help` output (to hear from people).

Thanks, Claudio! There are still many places where this list looks extremely weird to me, and will need to be corrected.

> --[no-]crack-status        emit a status line whenever a password is cracked
> --progress-every=N         emit a status line every N seconds
> --restore[=NAME]           restore an interrupted session [called NAME]

`--restore` should be moved to before `--[no-]crack-status` here.

> --make-charset=FILE        make a charset file. It will be overwritten

Should be near the modes. Also, I think we should avoid starting a new sentence in the one-line option description. FWIW, in core this was:

```
--make-charset=FILE        make a charset, FILE will be overwritten
```

> --fork=N                   fork N processes

Should be near `--node`, or the other way around.

> --fix-state-delay=N        performance tweak, see doc/OPTIONS

Should be dropped.

Several other things look moderately weird to me as well, but I don't immediately have suggested changes for them.

**Comment by magnumripper** [comment]
I assume @claudioandre-br catched all missing/added lines now so I didn't look for them. I think this is good now, after the mentioned stuff is fixed.

> > Final `--help` output (to hear from people).
> 
> Thanks, Claudio! There are still many places where this list looks extremely weird to me, and will need to be corrected.
> 
> > --[no-]crack-status        emit a status line whenever a password is cracked
> > --progress-every=N         emit a status line every N seconds
> > --restore[=NAME]           restore an interrupted session [called NAME]
> 
> `--restore` should be moved to before `--[no-]crack-status` here.

This could be accomplished by moving the other two way down to the alphabetical stuff. They are both "rarely used".

> FWIW, in core this was:
> ```
> --make-charset=FILE        make a charset, FILE will be overwritten
> ```

I think we should use the exact same phrasing.

**Comment by magnumripper** [comment]
> `--prince-elem-cnt-max=[N]`

This is wrong (in current code as well), please remove the brackets!

EDIT no wait, it should be like this!
```
--prince-elem-cnt-max=[-]N
```
Optional minus sign (we use that syntax elsewhere)

**Comment by magnumripper** [comment]
I just realized we have a number of descriptions starting with a capital letter, although most (and core) does not. Maybe that's for a separate PR. Personally I think a capital letter adds a little readability to the wall of text.

**Comment by magnumripper** [comment]
#4618 added two options, this PR need to be rebased on that.
```diff
@@ -297,6 +299,8 @@ JOHN_USAGE_REGEX \
 "--subsets-min-diff=N       Minimum unique characters in subset\n" \
 "--subsets-max-diff=[-]N    Maximum unique characters in subset (negative N is\n" \
 "                           relative to word length)\n" \
+"--subsets-prefer-short     Prefer shorter candidates over smaller subsets\n" \
+"--subsets-prefer-small     Prefer smaller subsets over shorter candidates\n" \
 "--stdout[=LENGTH]          just output candidate passwords [cut at LENGTH]\n" \
 "--restore[=NAME]           restore an interrupted session [called NAME]\n" \
 "--session=NAME             give a new session the NAME\n" \
```

**Comment by wsandin** [comment]
I made an attempt to update the --help list based on above feedback. I'm not sure if it's worth moving `--stdout` grouped with the logging and verbosity options? Happy to hear any additional feedback, and then update the `doc/OPTIONS` file.

@magnumripper I'd gladly add a capital letter for each opt description in this MR. It's a small and relatively easy change. So if no one else opposes, I can add that as last thing after squashing these commits, once we're all agree about the `--help` ordering.

**Current output**:
```
--help                     print usage summary
--single[=SECTION[,..]]    "single crack" mode, using default or named rules
--single=:rule[,..]        same, using "immediate" rule(s)
--single-seed=WORD[,WORD]  add static seed word(s) for all salts in single mode
--single-wordlist=FILE     *short* wordlist with static seed words/morphemes
--single-user-seed=FILE    wordlist with seeds per username (user:password[s]
                           format)
--single-pair-max=N        override max. number of word pairs generated (6)
--no-single-pair           disable single word pair generation
--[no-]single-retest-guess override config for SingleRetestGuess
--wordlist[=FILE] --stdin  wordlist mode, read words from FILE or stdin
                  --pipe   like --stdin, but bulk reads, and allows rules
--rules[=SECTION[,..]]     enable word mangling rules (for wordlist or PRINCE
                           modes), using default or named rules
--rules=:rule[;..]]        same, using "immediate" rule(s)
--rules-stack=SECTION[,..] stacked rules, applied after regular rules or to
                           modes that otherwise don't support rules
--rules-stack=:rule[;..]   same, using "immediate" rule(s)
--rules-skip-nop           skip any NOP ":" rules (you already ran w/o rules)
--loopback[=FILE]          like --wordlist, but extract words from a .pot file
--mem-file-size=SIZE       size threshold for wordlist preload (default 2048 MB)
--dupe-suppression         suppress all dupes in wordlist (and force preload)
--incremental[=MODE]       "incremental" mode [using section MODE]
--incremental-charcount=N  override CharCount for incremental mode
--external=MODE            external mode or word filter
--mask[=MASK]              mask mode using MASK (or default from john.conf)
--markov[=OPTIONS]         "Markov" mode (see doc/MARKOV)
--mkv-stats=FILE           "markov" stats file (see doc/MARKOV)
--prince[=FILE]            PRINCE mode, read words from FILE
--prince-loopback[=FILE]   fetch words from a .pot file
--prince-elem-cnt-min=N    minimum number of elements per chain (1)
--prince-elem-cnt-max=[N]  maximum number of elements per chain (negative N is
                           relative to word length) (8)
--prince-skip=N            initial skip
--prince-limit=N           limit number of candidates generated
--prince-wl-dist-len       calculate length distribution from wordlist
--prince-wl-max=N          load only N words from input wordlist
--prince-case-permute      permute case of first letter
--prince-mmap              memory-map infile (not available with case permute)
--prince-keyspace          just show total keyspace that would be produced
                           (disregarding skip and limit)
--subsets[=CHARSET]        "subsets" mode (see doc/SUBSETS)
--subsets-required=N       The N first characters of "subsets" charset are
                           the "required set" (see doc/SUBSETS)
--subsets-min-diff=N       Minimum unique characters in subset
--subsets-max-diff=[-]N    Maximum unique characters in subset (negative N is
--make-charset=FILE        make a charset, FILE will be overwritten
                           relative to word length)
--stdout[=LENGTH]          just output candidate passwords [cut at LENGTH]
--session=NAME             give a new session the NAME
--status[=NAME]            print status of a session [called NAME]
--restore[=NAME]           restore an interrupted session [called NAME]
--[no-]crack-status        emit a status line whenever a password is cracked
--progress-every=N         emit a status line every N seconds
--show[=left]              show cracked passwords [if =left, then uncracked]
--show=formats             show information about hashes in a file (JSON)
--show=invalid             show lines that are not valid for selected format(s)
--test[=TIME]              run tests and benchmarks for TIME seconds each
                           (if TIME is explicitly 0, test w/o benchmark)
--stress-test[=TIME]       loop self tests forever
--test-full=LEVEL          run more thorough self-tests
--no-mask                  used with --test for alternate benchmark w/o mask
--skip-self-tests          skip self tests
--users=[-]LOGIN|UID[,..]  [do not] load this (these) user(s) only
--groups=[-]GID[,..]       load users [not] of this (these) group(s) only
--shells=[-]SHELL[,..]     load users with[out] this (these) shell(s) only
--salts=[-]COUNT[:MAX]     load salts with[out] COUNT [to MAX] hashes, or
--salts=#M[-N]             load M [to N] most populated salts
--costs=[-]C[:M][,...]     load salts with[out] cost value Cn [to Mn]. For
                           tunable cost parameters, see doc/OPTIONS
--save-memory=LEVEL        enable memory saving, at LEVEL 1..3
--fork=N                   fork N processes
--node=MIN[-MAX]/TOTAL     this node's number range out of TOTAL count
--log-stderr               log to screen instead of file
--verbosity=N              change verbosity (1-5 or 6 for debug, default 3)
--no-log                   disables creation and writing to john.log file
--bare-always-valid=Y      treat bare hashes as valid (Y/N)
--config=FILE              use FILE instead of john.conf or john.ini
--encoding=NAME            input encoding (eg. UTF-8, ISO-8859-1). See also
                           doc/ENCODINGS.
--input-encoding=NAME      input encoding (alias for --encoding)
--target-encoding=NAME     output encoding (used by format, see doc/ENCODINGS)
--force-tty                set up terminal for reading keystrokes even if we're
                           not the foreground process
--field-separator-char=C   use 'C' instead of the ':' in input and pot files
--fix-state-delay=N        performance tweak, see doc/OPTIONS
--internal-codepage=NAME   codepage used in rules/masks (see doc/ENCODINGS)
--[no-]keep-guessing       try finding plaintext collisions
--list=WHAT                list capabilities, see --list=help or doc/OPTIONS
--length=N                 shortcut for --min-len=N --max-len=N
--min-length=N             request a minimum candidate length in bytes
--max-length=N             request a maximum candidate length in bytes
--max-candidates=[-]N      gracefully exit after this many candidates tried.
                           (if negative, reset count on each crack)
--max-run-time=[-]N        gracefully exit after this many seconds (if negative,
                           reset timer on each crack)
--mkpc=N                   request a lower max. keys per crypt
--no-loader-dupecheck      disable the dupe checking when loading hashes
--pot=NAME                 pot file to use
--regen-lost-salts=N       brute force unknown salts (see doc/OPTIONS)
--reject-printable         reject printable binaries
--tune=HOW                 tuning options (auto/report/N)
--subformat=FORMAT         pick a benchmark format for --format=crypt
--format=[NAME|CLASS][,..] force hash of type NAME. The supported formats can
                           be seen with --list=formats and --list=subformats.
                           Valid classes: dynamic, cpu, opencl, ztex, mask, omp,
                           all, enabled, disabled.
```

**Comment by magnumripper** [comment]
I think you still need to rebase upon latest code. It might work fine anyway but I'm not sure it'll end up right.

As for capitalizing stuff I'm all for that unless @solardiz prefers you to lowercase the others instead.

**Comment by wsandin** [comment]
> I think you still need to rebase upon latest code. It might work fine anyway but I'm not sure it'll end up right.
> 
> As for capitalizing stuff I'm all for that unless @solardiz prefers you to lowercase the others instead.

@magnumripper I see that I managed to slash the 2 line description for `--subsets-max-diff=` while rebasing this (fixed now). But the rebase should already be done. I fetched the origin repos, merged `upstream/bleeding-jumbo` into my `origin/bleeding-jumbo`, and then `rebased/merged` with the local `bleeding-jumbo` repo before pushing last nights changes. The two new options should be present in `options.c`.

I'll wait for @solardiz and @claudioandre-br feedback re. whether to capitalize the description or not.

**Comment by solardiz** [comment]
> whether to capitalize the description or not.

I'm OK with capitalizing them, primarily to visually separate the continuation lines (which won't be capitalized).

(All descriptions in core fit on one line, which isn't the case in jumbo.)

**Comment by magnumripper** [comment]
> the rebase should already be done. I fetched the origin repos, merged `upstream/bleeding-jumbo` into my `origin/bleeding-jumbo`, and then `rebased/merged` with the local `bleeding-jumbo` repo before pushing last nights changes. The two new options should be present in `options.c`.

Yet they are not, did you check? When I wrote that I only judged from your ["Current output"](https://github.com/openwall/john/pull/4572#issuecomment-798947645) but looking at current diff right now they are not there either.

**Comment by magnumripper** [comment]
> looking at current diff right now they are not there either.

Now I see them in current diff (I've had similar strange lags from GitHub recently). It does look right!

**Comment by magnumripper** [comment]
I did some checks and decided to merge this now before it goes astray again 😆

**Comment by magnumripper** [comment]
Thanks @wsandin!

**Comment by magnumripper** [comment]
Let's make the capitalizing a separate PR. Very trivial. The quicker you get away with it, the less chances of conflicts.

**Comment by magnumripper** [comment]
Only now I realized I didn't gave you time to squash. Sorry @solardiz but I felt this was too time consuming to let go on 😆 and good for you @wsandin because it'll make your statistics look better 😎

**Comment by solardiz** [comment]
I think this wasn't ready for merge, and we definitely should have squashed, but whatever. Thanks everyone!

**Comment by wsandin** [comment]
Thanks everyone involved. :pray: I also missed to update `doc/OPTIONS` before merge - I can create a separate PR for that though and tag it inside the same original issue.

Maybe we should consider the GitHub issue itself a work in progress - and move the discussion back there. Once we have agreed about the last changes we want in place for the release, I can create a new PR with exactly that. 

Re. squashing of commits, I have no good suggestions unless we're willing to temporarily revert the merge. I think I could have configured it to auto-squash here in GitHub (will do next time), but I usually do it manually right before Merge for any fixes/changes that's been requested during the review for easier tracking.

**Pull request merged**
